### ISSUE TITLE
"ValueError: Invalid Terraform resource path" running migrate4.py

### ISSUE THREAD

**Issue by david-walker-gfs** [terraform-google-modules/terraform-google-project-factory#424]
When upgrading from 3.2 to 4.0 of the project factory, I encountered the following issue. 

```
$ ./migrate4.py
Traceback (most recent call last):
  File "./migrate4.py", line 349, in <module>
    main(sys.argv)
  File "./migrate4.py", line 329, in main
    migrate(state, dryrun=args.dryrun, verbose=args.verbose)
  File "./migrate4.py", line 287, in migrate
    for path in read_state(statefile)
  File "./migrate4.py", line 287, in <listcomp>
    for path in read_state(statefile)
  File "./migrate4.py", line 175, in from_path
    "Invalid Terraform resource path {!r}".format(path))
ValueError: Invalid Terraform resource path 'google_logging_metric.deleted_service_account_log_metrics["datalake-core-dev"]'
```

Definition for the `google_logging_metric` resource is:

```
resource "google_logging_metric" "deleted_service_account_log_metrics" {
  for_each    = toset(local.monitored_projects)
  name        = "${each.key}/service_account_deleted"
  description = "Service Account has been deleted"
  project     = each.key
  filter      = "resource.type=service_account logName=\"projects/${each.key}/logs/cloudaudit.googleapis.com%2Factivity\" protoPayload.methodName=\"google.iam.admin.v1.DeleteServiceAccount\""

  metric_descriptor {
    metric_kind = "DELTA"
    value_type  = "INT64"
  }
}
```

The local is simple a list of projects.
```
  monitored_projects = [
...
    "datalake-core-${terraform.workspace}"
..
  ]
```

I've seen this before on another project when I previously tried to upgrade but never got around to look at it further. The similarities occur whenever there's quotations in the resource name, usually as a result of iterating over a list to create resources.

Obviously, the regex match is not expecting quotes in the resource name and that the reason for the error but I'm not sure to go about fixing this.
```
        if re.match(r'\A[\w.\[\]-]+\Z', path) is None:
            raise ValueError(
                "Invalid Terraform resource path {!r}".format(path))
```

I'm unable to upgrade the project factory as a result. Appreciate any help.

**Comment by david-walker-gfs** [comment]
Upon experimenting, it looks like if I just remove the check, it works. I am not sure if the implications of doing so though. Think perhaps the regex just needs to be updated to allow quoted resources.

**Comment by david-walker-gfs** [comment]
Another issue I've encountered is that the script fails if when running a normal apply, a tfvars file is expected. 

```
$ ./migrate4.py
---- Migrating the following project-factory modules:
-- module.project.module.project-factory
Removed module.project.module.project-factory.google_project_service.project_services[0]
Successfully removed 1 resource instance(s).
module.project.module.project-factory.module.project_services.google_project_service.project_services["compute.googleapis.com"]: Importing from ID "datalake-process-dev/compute.googleapis.com"...
module.project.module.project-factory.module.project_services.google_project_service.project_services["compute.googleapis.com"]: Import prepared!
  Prepared google_project_service for import
module.project.module.project-factory.module.project_services.google_project_service.project_services["compute.googleapis.com"]: Refreshing state... [id=datalake-process-dev/compute.googleapis.com]

Error: Invalid index

  on /git/datalake-terraform/env/datalake-process/variables.tf line 217, in locals:
 217:     metastore_pw            = var.metastore_pw[0]
    |----------------
    | var.metastore_pw is empty list of dynamic

The given key does not identify an element in this collection value.

Traceback (most recent call last):
  File "./migrate4.py", line 349, in <module>
    main(sys.argv)
  File "./migrate4.py", line 329, in main
    migrate(state, dryrun=args.dryrun, verbose=args.verbose)
  File "./migrate4.py", line 316, in migrate
    subprocess.run(argv, check=True, encoding='utf-8')
  File "/usr/local/lib/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['terraform', 'import', 'module.project.module.project-factory.module.project_services.google_project_service.project_services["compute.googleapis.com"]', 'datalake-process-dev/compute.googleapis.com']' returned non-zero exit status 1.
```

**Comment by morgante** [comment]
This error looks like it's coming from unrelated config (ie. elsewhere in your Terraform config). I'd suggest commenting out any other config while you run the import commands.

The migration script is really just meant as a helper, if you run it in with the `--dryrun` command it will output the required `terraform state` commands to migrate. You can then work through those one-by-one to fix any issues in your set up.

**Comment by david-walker-gfs** [comment]
Thanks morgante. I've managed to push my upgrade through by commenting out the validation as mentioned and then to get over the second issue with tfvars, I added a default value for the variable while it did the migrate. 

Perhaps this will help others who encounter issues with the upgrade script in future. I wasn't aware it was only supposed to be a helper. I guess the dryrun flag would help alright.

Feel free to close this off.
